src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferHashDataOverComputeHash.cs

```diff
@@ -0,0 +1,339 @@
+            CA1849,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            description: s_localizableStringDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(StringRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var compilation = context.Compilation;
+
+            if (!compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemSecurityCryptographyHashAlgorithm, out var hashAlgoBaseType) ||
+                !compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemSecurityCryptographySHA256, out var sha256Type))
+            {
+                return;
+            }
+
+            var byteType = context.Compilation.GetSpecialType(SpecialType.System_Byte);
+            if (byteType is null)
+            {
+                return;
+            }
+
+            var byteArrayParameter = ParameterInfo.GetParameterInfo(byteType, isArray: true, arrayRank: 1);
+
+            // method introduced in .NET 5.0
+            var hashDataMethodType = sha256Type.GetMembers(HashDataMethodName).OfType<IMethodSymbol>().GetFirstOrDefaultMemberWithParameterInfos(byteArrayParameter);
+            if (hashDataMethodType is null)
+            {
+                return;
+            }
+
+            var computeHashMethodBaseType = hashAlgoBaseType.GetMembers(ComputeHashMethodName).OfType<IMethodSymbol>().GetFirstOrDefaultMemberWithParameterInfos(byteArrayParameter);
+            if (computeHashMethodBaseType is null)
+            {
+                return;
+            }
+
+            context.RegisterOperationBlockStartAction(OnOperationBlockStart);
+            return;
+
+            void OnOperationBlockStart(OperationBlockStartAnalysisContext context)
+            {
+                // Patterns we are looking for:
+                // Pattern #1                                               Pattern #2
+                // var obj = #HashCreation#;              or                #HashCreation#.CompuateHash(buffer);
+                // ...
+                // obj.CompuateHash(buffer); //x1
+
+                // The core search logic is split into 3 groups
+                // 1. Scan all invocations:
+                //  a. if it is a Hash Create static method, store its local symbol + declaration operation and type symbols
+                //  b. if it is HashAlgorithm.ComputeHash
```

Comment:
Don't we want to cover/replace
`public bool TryComputeHash(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesWritten)` with 
`public static bool TryHashData(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesWritten)`?

Reply:
hmm, there's also `ComputeHash(Byte[], Int32, Int32)`

Reply:
> hmm, there's also ComputeHash(Byte[], Int32, Int32)

I don't see a direct replacement for that, probably safe to leave it.

@bartonjs is is safe to replace 
`public bool TryComputeHash(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesWritten)` with 
`public static bool TryHashData(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesWritten)`?

Reply:
`ComputeHash(byte[] arr, int offset, int length)` would map to `HashData(arr.AsSpan(offset, length))`

And, yeah, instance `TryComputeHash` can go to static `TryHashData`.

Reply:
I've done this but I have to get more test written

Reply:
To simplify the analyzer, my preference would be that it fire only on this simple pattern:

1. The analyzer detects a fresh instantiation of the algorithm (via Create or whatever other mechanism is appropriate); and
2. The *only* thing done with this instance before disposal is one or more calls to ComputeHash.

If anything else at all is done with this instance - passed as an argument to a method, returned out of the method, somebody makes a copy of the local, somebody calls GetType on it, *anything* - the analyzer should not fire. Disposal is optional and is allowed as long as it's the very last thing done with the instance.

This means we don't have to worry about the TransformBlock / TransformFinalBlock cases, and we don't have to worry about other cases where somebody may have mutated the state of the internal object to some unknown value. Even though this analyzer isn't 100% comprehensive, it should be sufficient to capture the 95% usage case.

If we want to be smarter about fixing TransformBlock / TransformFinalBlock, we can save that for a future analyzer.

Reply:
Yes, at the moment, it checks if the instance is only calling `ComputeHash`.

There isn't a check on an explicit call to dispose atm.

Reply:
What's the recommended approach in a roslyn analyzer to detect an invocation is the last invocation?

Reply:
> What's the recommended approach in a roslyn analyzer to detect an invocation is the last invocation?

Don't know if there is a way to detect that, not sure why/for what you need that, i think whatever you are trying to do you can do that in the `OperationBlockEndAction` as that would run after all invocations done

Reply:
There's no check on Dispose being the last method invocation in the logic at the moment.

Reply:
> There's no check on Dispose being the last method invocation in the logic at the moment.

I don't think you need to check if the disposal is the last invocation (if it is not the last invocation then it would throw Disposed exception anyways), i think the only thing matter is `Dispose` is allowed to be called on the instance other than `ComputeHash` which can be done in `OperationBlockEndAction`

Reply:
ok, sound good